In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.dummy import DummyClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import MultinomialNB
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.decomposition import TruncatedSVD, PCA
from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score, confusion_matrix
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer, HashingVectorizer
from sklearn import datasets
from sklearn.model_selection import GridSearchCV
from nltk.corpus import names, movie_reviews
import nltk
import numpy as np
import pandas as pd
nltk.download('all')

[nltk_data] Downloading collection 'all'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/abc.zip.
[nltk_data]    | Downloading package alpino to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/alpino.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping taggers/averaged_perceptron_tagger.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger_ru to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping
[nltk_data]    |       taggers/averaged_perceptron_tagger_ru.zip.
[nltk_data]    | Downloading package basque_grammars to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping grammars/basque_grammars.zip.
[nltk_data]    | Downloading package bcp47 to /root/nltk_data...
[nltk_data]    | Downloading package biocreative_ppi to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   U

True

In [ ]:
twenty_users = datasets.fetch_20newsgroups()

In [ ]:
twenty_users.target_names

['alt.atheism',
 'comp.graphics',
 'comp.os.ms-windows.misc',
 'comp.sys.ibm.pc.hardware',
 'comp.sys.mac.hardware',
 'comp.windows.x',
 'misc.forsale',
 'rec.autos',
 'rec.motorcycles',
 'rec.sport.baseball',
 'rec.sport.hockey',
 'sci.crypt',
 'sci.electronics',
 'sci.med',
 'sci.space',
 'soc.religion.christian',
 'talk.politics.guns',
 'talk.politics.mideast',
 'talk.politics.misc',
 'talk.religion.misc']

In [ ]:
df = pd.DataFrame({'doc': twenty_users.data, 'target': twenty_users.target})
traindf, testdf = train_test_split(df)

In [ ]:
from nltk.tokenize import word_tokenize
distribution = nltk.FreqDist(word for doc in traindf['doc'] for word in word_tokenize(doc))

In [ ]:
total_words = len(traindf['doc'])
words975 = [word for word, freq in distribution.items() if freq / total_words >= 0.975 or freq / total_words <= 0.025]

In [ ]:
vectorizer = CountVectorizer(stop_words=words975)

In [ ]:
pipeline = Pipeline([
    ('vectorizer', vectorizer),
    ('classifier', RandomForestClassifier(n_estimators=50))
])
pipeline.fit(traindf['doc'], traindf['target'])

/usr/local/lib/python3.10/dist-packages/sklearn/feature_extraction/text.py:409: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['00000', '000000', '00000000b', '00000001b', '00000010b', '00000011b', '000000e5', '00000100b', '00000101b', '00000110b', '00000111b', '000007', '00000ee5', '00001000b', '00001001b', '00001010b', '00001011b', '000010af', '00001100b', '00001101b', '00001110b', '00001111b', '000042', '000062david42', '000094', '0000vec', '00010000b', '00010001b', '00010010b', '00010011b', '00010100b', '00010101b', '00010110b', '00010111b', '00011000b', '00011001b', '00011010b', '00011011b', '00011100b', '00011101b', '00011110b', '00011111b', '00014', '000152', '0001mpc', '0001x7c', '000246', '000256', '000359', '0004', '00041032', '000413', '00044513', '0005', '0005111312na1em', '0005111312na3em', '000531', '000601', '0007', '000710', '0007259d', '00072a27', '000773b9', '00077bbe', '000851', '0009', '00090711'

Pipeline(steps=[('vectorizer',
                 CountVectorizer(stop_words=['From', ':', 'jrwaters', '@',
                                             'eos.ncsu.edu', '(', 'JACK',
                                             'ROGERS', 'WATERS', ')', 'Subject',
                                             'and', 'it', 'to', '.', 'Carolina',
                                             ',', 'Project', 'Eos', 'Lines',
                                             '74', '<',
                                             '1993Apr20.204831.19788',
                                             'rtsg.mot.com', '>', 'svoboda',
                                             'Svoboda', '13516',
                                             'news.duke.edu', 'infante', ...])),
                ('classifier', RandomForestClassifier(n_estimators=50))])

In [ ]:
training_accuracy = accuracy_score(traindf['target'], pipeline.predict(traindf['doc']))
training_accuracy

0.9998821449616971

In [ ]:
testing_accuracy = accuracy_score(testdf['target'], pipeline.predict(testdf['doc']))
testing_accuracy

0.7317073170731707

The training accuracy is 99% which means the data fits perfectly to the training data, meaning the model memorized the data and is overfitting.

The testing accuracy is 73%, which is almost same as when the dimensionality was decreased.